libs/execution_context/ebpf_maps.c

```diff
@@ -944,6 +985,98 @@ _find_lpm_trie_map_entry(_In_ ebpf_core_map_t* map, _In_ const uint8_t* key)
     return best_value;
 }
 
+static ebpf_core_map_t*
+_create_queue_map(_In_ const ebpf_map_definition_in_memory_t* map_definition)
+{
+    size_t queue_map_size =
+        EBPF_OFFSET_OF(ebpf_core_queue_map_t, slots) + map_definition->max_entries * sizeof(uint8_t*);
+    return _create_array_map_with_map_struct_size(queue_map_size, map_definition);
+}
+
+static void
+_delete_queue_map(_In_ ebpf_core_map_t* map)
+{
+    ebpf_core_queue_map_t* queue_map = EBPF_FROM_FIELD(ebpf_core_queue_map_t, core_map, map);
+
+    for (size_t i = 0; i < queue_map->core_map.ebpf_map_definition.max_entries; i++) {
+        ebpf_epoch_free(queue_map->slots[i]);
+        queue_map->slots[i] = NULL;
+    }
+    ebpf_epoch_free(queue_map);
+}
+
+static uint8_t*
+_peek_or_pop_queue_map_entry(ebpf_core_queue_map_t* queue_map, bool peek)
+{
+    uint8_t* value = NULL;
+    if (queue_map->start == queue_map->end) {
+        return NULL;
+    }
+
+    ebpf_lock_state_t state = ebpf_lock_lock(&queue_map->core_map.lock);
+    value = queue_map->slots[queue_map->start];
+    if (!peek) {
+        ebpf_epoch_free(queue_map->slots[queue_map->start]);
+        queue_map->slots[queue_map->start] = NULL;
+        queue_map->start = (queue_map->start + 1) % queue_map->core_map.ebpf_map_definition.max_entries;
+    }
+    ebpf_lock_unlock(&queue_map->core_map.lock, state);
+    return value;
+}
+
+static uint8_t*
+_find_queue_map_entry(_In_ ebpf_core_map_t* map, _In_ const uint8_t* key)
+{
+    if (!map || key)
+        return NULL;
+
+    ebpf_core_queue_map_t* queue_map = EBPF_FROM_FIELD(ebpf_core_queue_map_t, core_map, map);
+    return _peek_or_pop_queue_map_entry(queue_map, true);
+}
+
+static ebpf_result_t
+_update_queue_map_entry(
+    _In_ ebpf_core_map_t* map, _In_ const uint8_t* key, _In_opt_ const uint8_t* data, ebpf_map_option_t option)
+{
+    ebpf_result_t result;
+    if (!map || key || (option == EBPF_NOEXIST))
+        return EBPF_INVALID_ARGUMENT;
+
+    ebpf_core_queue_map_t* queue_map = EBPF_FROM_FIELD(ebpf_core_queue_map_t, core_map, map);
+    ebpf_lock_state_t state = ebpf_lock_lock(&queue_map->core_map.lock);
+
+    // Check for queue full
+    if (queue_map->start == ((queue_map->end + 1) % queue_map->core_map.ebpf_map_definition.max_entries)) {
+        result = EBPF_INVALID_ARGUMENT;
+        goto Done;
+    }
+
+    uint8_t* value = ebpf_epoch_allocate(queue_map->core_map.ebpf_map_definition.value_size);
+    if (!value) {
+        result = EBPF_NO_MEMORY;
+        goto Done;
+    }
+    memcpy(value, data, queue_map->core_map.ebpf_map_definition.value_size);
+    queue_map->slots[queue_map->end] = value;
+
+    queue_map->end = (queue_map->end + 1) % queue_map->core_map.ebpf_map_definition.max_entries;
+    result = EBPF_SUCCESS;
+
+Done:
+    ebpf_lock_unlock(&queue_map->core_map.lock, state);
+    return result;
+}
+
+uint8_t*
+_find_and_delete_queue_map_entry(_In_ ebpf_core_map_t* map, _In_ const uint8_t* key)
```

Comment:
```suggestion
_find_and_delete_queue_map_entry(_In_ ebpf_core_map_t* map, _In_opt_ const uint8_t* key)
```
Since the main value of key is NULL (per 1073), it should be `_In_opt_`